File: ios/EasyUnlocker/Crypto/VaultCrypto.swift
import CryptoKit
import Foundation
import Security

/// 保险库落盘形状（vault.eu1），与安卓端 VaultFile 字段一致：
/// argon2id(恢复码, salt, ops, memKiB) → AES-256-GCM(AAD="easy-unlocker-vault")。
struct VaultFile: Codable {
    var v: Int = 1
    var kdf: String = "argon2id"
    var salt: String
    var nonce: String
    var ciphertext: String
    var ops: Int = 3
    var memKiB: Int = 64 * 1024
}

enum VaultCrypto {
    static let aad = Data("easy-unlocker-vault".utf8)

    static func newRecoveryCode() -> String {
        var raw = [UInt8](repeating: 0, count: 16)
        _ = SecRandomCopyBytes(kSecRandomDefault, 16, &raw)
        return raw.map { String(format: "%02x", $0) }.joined()
            .chunked(into: 4).joined(separator: "-")
    }

    /// "xxxx-xxxx-…" → 16 字节；长度不对直接抛错。
    static func normalizeRecovery(_ code: String) throws -> Data {
        let hex = code.lowercased()
            .replacingOccurrences(of: "-", with: "")
            .replacingOccurrences(of: " ", with: "")
        guard hex.count == 32 else { throw EuError.badInput("恢复码长度不对") }
        var out = Data()
        var i = hex.startIndex
        while i < hex.endIndex {
            let j = hex.index(i, offsetBy: 2)
            guard let b = UInt8(hex[i..<j], radix: 16) else {
                throw EuError.badInput("恢复码里有非法字符")
            }
            out.append(b)
            i = j
        }
        return out
    }

    static func deriveKey(recovery: Data, salt: Data, ops: Int, memKiB: Int) throws -> Data {
        try Argon2.deriveKey(password: recovery, salt: salt, ops: UInt32(ops), memKiB: UInt32(memKiB))
    }

    /// AES-256-GCM：返回（nonce, 密文+tag）。tag 128bit 与安卓 GCMBlockCipher 一致。
    static func encryptVault(key: Data, plaintext: Data) throws -> (nonce: Data, ciphertext: Data) {
        let nonce = AES.GCM.Nonce()
        let box = try AES.GCM.seal(plaintext, using: SymmetricKey(data: key), nonce: nonce, authenticating: aad)
        return (Data(nonce), box.ciphertext + box.tag)
    }

    static func decryptVault(key: Data, nonce: Data, ciphertext: Data) throws -> Data {
        guard ciphertext.count >= 16 else { throw EuError.badInput("密文太短") }
        let ct = ciphertext.prefix(ciphertext.count - 16)
        let tag = ciphertext.suffix(16)
        let box = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: nonce), ciphertext: ct, tag: tag)
        return try AES.GCM.open(box, using: SymmetricKey(data: key), authenticating: aad)
    }

    static func b64(_ raw: Data) -> String { raw.base64EncodedString() }
    static func unb64(_ value: String) -> Data? { Data(base64Encoded: value) }
}

enum EuError: LocalizedError {
    case badInput(String)
    case locked
    case crypto(String)
    case network(String)
    var errorDescription: String? {
        switch self {
        case .badInput(let m), .crypto(let m), .network(let m): return m
        case .locked: return "locked"
        }
    }
}

extension String {
    func chunked(into n: Int) -> [String] {
        var out: [String] = []
        var i = startIndex
        while i < endIndex {
            let j = index(i, offsetBy: n, limitedBy: endIndex) ?? endIndex
            out.append(String(self[i..<j]))
            i = j
        }
        return out
    }
}

File: android/app/src/main/java/io/github/cyancity/easyunlocker/crypto/VaultCrypto.kt
package io.github.cyancity.easyunlocker.crypto

import android.util.Base64
import org.bouncycastle.crypto.generators.Argon2BytesGenerator
import org.bouncycastle.crypto.params.Argon2Parameters
import org.bouncycastle.crypto.engines.AESEngine
import org.bouncycastle.crypto.modes.GCMBlockCipher
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.KeyParameter
import java.security.SecureRandom

data class VaultFile(
    val v: Int = 1,
    val kdf: String = "argon2id",
    val salt: String,
    val nonce: String,
    val ciphertext: String,
    val ops: Int = 3,
    val memKiB: Int = 64 * 1024,
)

object VaultCrypto {
    fun newRecoveryCode(): String {
        val raw = ByteArray(16)
        SecureRandom().nextBytes(raw)
        return raw.joinToString("") { "%02x".format(it) }.chunked(4).joinToString("-")
    }

    fun normalizeRecovery(code: String): ByteArray {
        val hex = code.lowercase().replace("-", "").replace(" ", "")
        require(hex.length == 32) { "恢复码长度不对" }
        return hex.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }

    fun deriveKey(recovery: ByteArray, salt: ByteArray, ops: Int, memKiB: Int): ByteArray {
        val params = Argon2Parameters.Builder(Argon2Parameters.ARGON2_id)
            .withSalt(salt)
            .withIterations(ops)
            .withMemoryAsKB(memKiB)
            .withParallelism(1)
            .build()
        val gen = Argon2BytesGenerator()
        gen.init(params)
        val out = ByteArray(32)
        gen.generateBytes(recovery, out)
        return out
    }

    fun encryptVault(key: ByteArray, plaintext: ByteArray): Pair<ByteArray, ByteArray> {
        val nonce = ByteArray(12).also { SecureRandom().nextBytes(it) }
        val cipher = GCMBlockCipher.newInstance(AESEngine.newInstance())
        cipher.init(true, AEADParameters(KeyParameter(key), 128, nonce, "easy-unlocker-vault".toByteArray()))
        val out = ByteArray(cipher.getOutputSize(plaintext.size))
        val n = cipher.processBytes(plaintext, 0, plaintext.size, out, 0)
        cipher.doFinal(out, n)
        return nonce to out
    }

    fun decryptVault(key: ByteArray, nonce: ByteArray, ciphertext: ByteArray): ByteArray {
        val cipher = GCMBlockCipher.newInstance(AESEngine.newInstance())
        cipher.init(false, AEADParameters(KeyParameter(key), 128, nonce, "easy-unlocker-vault".toByteArray()))
        val out = ByteArray(cipher.getOutputSize(ciphertext.size))
        val n = cipher.processBytes(ciphertext, 0, ciphertext.size, out, 0)
        cipher.doFinal(out, n)
        return out
    }

    fun b64(raw: ByteArray): String = Base64.encodeToString(raw, Base64.NO_WRAP)
    fun unb64(value: String): ByteArray = Base64.decode(value, Base64.NO_WRAP)
}
